plugins/ingest/src/test/resources/rest-api-spec/test/ingest/80_simulate.yaml

```diff
@@ -0,0 +1,274 @@
+                "_source": {
+                  "foo": "bar"
+                }
+              }
+            ]
+          }
+  - length: { error: 3 }
+  - match: { status: 400 }
+  - match: { error.type: "illegal_argument_exception" }
+  - match: { error.reason: "required property [pipeline] is missing" }
+
+---
+"Test simulate with verbose flag":
+  - do:
+      cluster.health:
+          wait_for_status: green
+
+  - do:
+      ingest.simulate:
+        verbose: true
+        body: >
+          {
+            "pipeline": {
+              "description": "_description",
+              "processors": [
+                {
+                  "mutate" : {
+                    "update" : {
+                      "field2" : "_value"
+                    }
+                  }
+                },
+                {
+                  "mutate" : {
+                    "update" : {
+                      "field3" : "third_val"
+                    }
+                  }
+                }
+              ]
+            },
+            "docs": [
+              {
+                "_index": "index",
+                "_type": "type",
+                "_id": "id",
+                "_source": {
+                  "foo": "bar"
+                }
+              }
+            ]
+          }
+  - length: { docs: 1 }
+  - length: { docs.0.processor_results: 2 }
+  - match: { docs.0.processor_results.0.processor_id: "processor[mutate]-0" }
+  - is_true: docs.0.processor_results.0.doc.modified
+  - length: { docs.0.processor_results.0.doc._source: 2 }
+  - match: { docs.0.processor_results.0.doc._source.foo: "bar" }
+  - match: { docs.0.processor_results.0.doc._source.field2: "_value" }
+  - length: { docs.0.processor_results.1.doc._source: 3 }
+  - match: { docs.0.processor_results.1.doc._source.foo: "bar" }
+  - match: { docs.0.processor_results.1.doc._source.field2: "_value" }
+  - match: { docs.0.processor_results.1.doc._source..field3: "third_val" }
+
+---
+"Test simulate with exception thrown":
+  - do:
+      cluster.health:
+          wait_for_status: green
+
+  - do:
+      ingest.simulate:
+        body: >
+          {
+            "pipeline": {
+              "description": "_description",
+              "processors": [
+                {
+                  "mutate" : {
+                    "uppercase" : ["foo"]
+                  }
+                }
+              ]
+            },
+            "docs": [
+              {
+                "_index": "index",
+                "_type": "type",
+                "_id": "id",
+                "_source": {
+                  "not_foo": "bar"
+                }
+              },
+              {
+                "_index": "index",
+                "_type": "type",
+                "_id": "id2",
+                "_source": {
+                  "foo": "bar"
+                }
+              }
+            ]
+          }
+  - length: { docs: 2 }
+  - match: { docs.0.error.type: "null_pointer_exception" }
```

Comment:
I didn't realize we would return NPE. Is that because a field is not there? In general I think we should return nicer errors as much as we can.